SDK Auth [do not merge] (#1679)
braze-inc/braze-docs: _docs/_developer_guide/platform_wide/sdk_authentication.md

---
nav_title: SDK Authentication
page_order: 5
hidden: true
description: "Verify the identity of SDK requests"
platform:
  - ios
  - android
  - web
---

<br>
{% alert important %}
This feature is in _beta_ and only available for Android, Web, and iOS SDKs.
{% endalert %}

# SDK Authentication

> SDK Authentication ensures that no one can impersonate or create Braze users using your SDK API Keys. Braze SDKs, like other client-side libraries, use a public API key to identify your app. To ensure that Braze only trusts requests from authenticated users, this feature allows _you_ to supply Braze with secure, cryptographic proof. After all, your server is the only one who would know whether a user is who they claim to be in your app (using cookies or other persistent session data).

Reviewer: I think it would be worth considering how much of this high level description to keep internal, because of what it advertises isn't currently "secure".

Reviewer 2: Agreed.

Author: ```suggestion
SDK Authentication allows you to supply cryptographic proof to SDK requests made on behalf of logged-in users. When  this feature is enabled in your app, Braze can (optionally) reject requests which are made without a valid JWT signature.
```

Author: @Bucimis @radixdev how's that?

When enabled, this feature will prevent unauthorized requests that use your app's SDK API Key for logged in users, including:
- Sending custom events, attributes, purchases, and session data
- Creating new users in your Braze App Group
- Updating standard user profile attributes
- Receiving or triggering messages

## Getting Started

There are four high-level steps to get started:

1. [Server-Side Integration][1] - Generate a public and private key-pair, and use your private key to create a JWT (_JSON Web Token_) for the current logged-in user.<br><br>

> 2. [SDK Integration][2] - Enable this feature in the Braze SDK and supply the JWT Token generated from your server.<br><br>

Reviewer: ```suggestion
2. [SDK Integration][2] - Enable this feature in the Braze SDK and request the JWT Token generated from your server.<br><br>
```

optional, but imo looking at it from the SDK standpoint, its a request

### JWT Libraries

To learn more about JSON Web Tokens, or to browse the [many open source libraries](https://jwt.io/#libraries-io) that simplify this signing process, check out [https://jwt.io](https://jwt.io).

## SDK Integration {#sdk-integration}

This feature is available as of the following [SDK versions]({{ site.baseurl }}/user_guide/engagement_tools/campaigns/ideas_and_strategies/new_features/#filtering-by-most-recent-app-versions):

{% sdk_min_versions web:3.4.0 ios:4.2.0 android:14.0.0 %}

### Enable this feature in the Braze SDK.

When this feature is enabled, the Braze SDK will append the current user's last known JWT to network requests made to Braze Servers.

{% alert note %}
Don't worry, initializing with this option alone won't impact data collection in any way, until you start [enforcing authentication](#braze-dashboard) within the Braze Dashboard.
{% endalert %}

{% tabs %}
{% tab Javascript %}
When calling `appboy.initialize`, set the optional `sdkAuthentication` property to `true`.
```javascript
appboy.initialize("YOUR-API-KEY-HERE", {
  baseUrl: "YOUR-SDK-ENDPOINT-HERE",
  sdkAuthentication: true,
});
```
{% endtab %}

> {% tab Java %}
> When configuring the Appboy instance, call `setIsSdkAuthenticationEnabled` to `true`.
> ```java
> AppboyConfig.Builder appboyConfigBuilder = new AppboyConfig.Builder()
>     .setIsSdkAuthenticationEnabled(true);
> Appboy.configure(this, appboyConfigBuilder.build());
> ```
> {% endtab %}

Reviewer 2: ```suggestion
{% tab Java %}
When configuring the Appboy instance, call `setIsSdkAuthenticationEnabled` to `true`.
```java
AppboyConfig.Builder appboyConfigBuilder = new AppboyConfig.Builder()
    .setIsSdkAuthenticationEnabled(true);
Appboy.configure(this, appboyConfigBuilder.build());
```
{% endtab %}
{% tab KOTLIN %}
When configuring the Appboy instance, call `setIsSdkAuthenticationEnabled` to `true`.
```kotlin
AppboyConfig.Builder appboyConfigBuilder = AppboyConfig.Builder()
    .setIsSdkAuthenticationEnabled(true)
Appboy.configure(this, appboyConfigBuilder.build())
```
{% endtab %}
```

        withLaunchOptions:launchOptions
        withAppboyOptions:@{ABKEnableSDKAuthenticationKey : @YES}];
```
{% endtab %}
{% tab Swift %}
```swift
Appboy.start(withApiKey: "YOUR-API-KEY",
                 in:application,
                 withLaunchOptions:launchOptions,
                 withAppboyOptions:[ ABKEnableSDKAuthenticationKey : true ])
```
{% endtab %}
{% endtabs %}

### Set the current user's JWT Token

Whenever your app calls the Braze `changeUser` method, also supply the JWT token that was [generated server-side][4].

You can also update the token to refresh the token mid-session for the current user.

{% alert note %}
Keep in mind that `changeUser` should only be called when the User ID has _actually changed_. You should not use this method as a way to update the signature if the user ID has not changed.
{% endalert %}

{% tabs %}
{% tab Javascript %}
Supply the JWT Token when calling `appboy.changeUser`:

```javascript
appboy.changeUser("NEW-USER-ID", "JWT-TOKEN-FROM-SERVER");
```

Or, when you have refreshed the user's token mid-session:

```javascript
appboy.setAuthenticationToken("NEW-JWT-TOKEN-FROM-SERVER");
```
{% endtab %}

> {% tab Java %}
> 
> Supply the JWT Token when calling `appboy.changeUser`:
> 
> ```java
> Appboy.getInstance(this).changeUser("NEW-USER-ID", "JWT-TOKEN-FROM-SERVER");
> ```
> 
> Or, when you have refreshed the user's token mid-session:
> 
> ```java
> Appboy.getInstance(this).setSdkAuthenticationSignature("NEW-JWT-TOKEN-FROM-SERVER");
> ```
> {% endtab %}

Reviewer 2: ```suggestion
{% tab Java %}

Supply the JWT Token when calling `appboy.changeUser`:

```java
Appboy.getInstance(this).changeUser("NEW-USER-ID", "JWT-TOKEN-FROM-SERVER");
```

Or, when you have refreshed the user's token mid-session:

```java
Appboy.getInstance(this).setSdkAuthenticationSignature("NEW-JWT-TOKEN-FROM-SERVER");
```
{% endtab %}
{% tab KOTLIN %}

Supply the JWT Token when calling `appboy.changeUser`:

```kotlin
Appboy.getInstance(this).changeUser("NEW-USER-ID", "JWT-TOKEN-FROM-SERVER")
```

Or, when you have refreshed the user's token mid-session:

```kotlin
Appboy.getInstance(this).setSdkAuthenticationSignature("NEW-JWT-TOKEN-FROM-SERVER")
```
{% endtab %}
```

#### What expiration should I use when generating JWT tokens? {#faq-expiration}

We recommend using the higher value of: average session duration, session cookie/token expiration, or the frequency at which your application would otherwise refresh the current user's profile.

#### What happens if a JWT expires in the middle of a user's session?

Should a user's token expire mid-session, the SDK has a [callback function][7] it will invoke to let your app know that a new JWT token is needed to continue sending data to Braze.

#### What happens if my server-side integration breaks and I can no longer create a JWT?

If your server is not able to provide JWT tokens or you notice some integration issue, you can always disable the feature in the Braze Dashboard.

Once disabled, any pending failed SDK requests will eventually be retried by the SDK and accepted by Braze.

#### Why does this feature use Public/Private keys instead of Shared Secrets?

When using Shared Secrets, anyone with access to that shared secret (i.e. the Braze Dashboard page) would be able to generate tokens and impersonate your end-users.

> Instead, we use Public/Private Keys so that not even Braze Employees (let alone your Dashboard users) have access to reveal your Private Keys.

Reviewer: ```suggestion
Instead, we use Public/Private Keys so that not even Braze Employees (let alone your Dashboard users) have access your Private Keys.
```

When this feature is enabled, the Braze SDK will append the current user's last known JWT to network requests made to Braze Servers.

{% alert note %}
Don't worry, initializing with this option alone won't impact data collection in any way, until you start [enforcing authentication](#braze-dashboard) within the Braze Dashboard.
{% endalert %}

{% tabs %}
{% tab Javascript %}
When calling `appboy.initialize`, set the optional `sdkAuthentication` property to `true`.
```javascript
appboy.initialize("YOUR-API-KEY-HERE", {
  baseUrl: "YOUR-SDK-ENDPOINT-HERE",
  sdkAuthentication: true,
});
```
{% endtab %}
{% tab Java %}
When configuring the Appboy instance, call `setIsSdkAuthenticationEnabled` to `true`.
```java
AppboyConfig.Builder appboyConfigBuilder = new AppboyConfig.Builder()
    .setIsSdkAuthenticationEnabled(true);
Appboy.configure(this, appboyConfigBuilder.build());
```
{% endtab %}
{% tab KOTLIN %}
When configuring the Appboy instance, call setIsSdkAuthenticationEnabled to true.
```kotlin
AppboyConfig.Builder appboyConfigBuilder = AppboyConfig.Builder()
    .setIsSdkAuthenticationEnabled(true)
Appboy.configure(this, appboyConfigBuilder.build())
```
{% endtab %}

> {% tab Objective-C %}

Author: ```suggestion
{% tab Objective-C %}
Set the `ABKEnableSDKAuthenticationKey` option set to `@YES`. Alternatively, this can be added to your `plist` options.

```

Author: @chshapiro @Bucimis can you help make this make sense :)

Reviewer 3: the key for plist is different - it's `EnableSDKAuthentication`

Reviewer 3: added a suggestion

> {% tab Objective-C %}

Reviewer 3: ```suggestion
{% tab Objective-C %}
To enable SDK Authentication, add the key `EnableSDKAuthentication` to the `Braze` dictionary in your `.plist` file and set it to true.

Alternatively, you can enable SDK Authentication when initializing the SDK:
```

```objc
[Appboy startWithApiKey:@"YOUR-API-KEY"
            inApplication:application
        withLaunchOptions:launchOptions

>         withAppboyOptions:@{ABKEnableSDKAuthenticationKey : @YES}];

Reviewer 3: This can also be set in the plist instead of in `appboyOptions` - I'm not sure where it would make sense to add that

```
{% endtab %}

> {% tab Swift %}

Author: ```suggestion
{% tab Swift %}
Set the `ABKEnableSDKAuthenticationKey` option set to `@YES`. Alternatively, this can be added to your `plist` options using `EnableSDKAuthentication`.

```

```swift

> Appboy.start(withApiKey: "YOUR-API-KEY",

Author: @chshapiro @Bucimis should we add the same comment above here? (plist)

Reviewer 3: yes same comment

```javascript
appboy.onSdkAuthenticationFailure(async (errorEvent) => {
  // TODO: check if the errorEvent user matches the currently logged-in user
  const updated_jwt = await getNewTokenSomehow(errorEvent);
  // TODO: optionally log to your error-reporting service
  appboy.setAuthenticationToken(updated_jwt);
});
```
{% endtab %}
{% tab Java %}
```java
Appboy.getInstance(this).subscribeToSdkAuthenticationFailures(errorEvent -> {
    // TODO: check if the errorEvent user matches the currently logged-in user
    String newToken = getNewTokenSomehow(errorEvent);
    // TODO: optionally log to your error-reporting service
    Appboy.getInstance(getContext()).setSdkAuthenticationSignature(newToken);
});
```
{% endtab %}
{% tab KOTLIN %}
```kotlin
Appboy.getInstance(this).subscribeToSdkAuthenticationFailures({ errorEvent: BrazeSdkAuthenticationErrorEvent ->
    // TODO: check if the errorEvent user matches the currently logged-in user
    val newToken: String = getNewTokenSomehow(errorEvent)
    // TODO: optionally log to your error-reporting service
    Appboy.getInstance(getContext()).setSdkAuthenticationSignature(newToken)
})
```
{% endtab %}
{% tab Objective-C %}
```objc

> [[Appboy sharedInstance] setSdkAuthenticationDelegate:delegate];

Reviewer 3: we should probably add more info here about the delegate - it's called `ABKSdkAuthenticationDelegate`. They need to create the delegate and implement its method `handleSdkAuthenticationError:` and then set the delegate using this method.
Once the code is live, we can only link to the delegate file in github

Author: sorry, i'm not sure what should be added here. would you be able to stake a stab here and below (or @Bucimis)?

> // Method to implement in delegate

Reviewer 3: might make sense to have this as a separate code snippet with a separate little intro

- (void)handleSdkAuthenticationError:(ABKSdkAuthenticationError *)errorEvent {
  // TODO: check if the errorEvent's user matches the currently logged-in user
  NSLog(@"Invalid SDK Authentication signature.");
  // TODO: optionally log to your error-reporting service
  NSString *newSignature = getNewSignatureSomehow(errorEvent);
  [[Appboy sharedInstance] setSdkAuthenticationSignature:newSignature];
}
```
{% endtab %}
{% tab Swift %}
```swift
Appboy.sharedInstance()?.setSdkAuthenticationDelegate(delegate)

// Method to implement in delegate

> func handle(_ authError: ABKSdkAuthenticationError?) {

Reviewer 3: Do you want to change `authError` to `errorEvent` here also?

> func handle(_ authError: ABKSdkAuthenticationError?) {

Author: ```suggestion
func handle(_ errorEvent: ABKSdkAuthenticationError?) {
```

        // TODO: check if the errorEvent's user matches the currently logged-in user
        print("Invalid SDK Authentication signature.")

>         let newSignature = getNewSignatureSomehow(authError)

Author: ```suggestion
        let newSignature = getNewSignatureSomehow(errorEvent)
```

{% alert note %}
Don't worry, initializing with this option alone won't impact data collection in any way, until you start [enforcing authentication](#braze-dashboard) within the Braze Dashboard.
{% endalert %}

{% tabs %}
{% tab Javascript %}
When calling `appboy.initialize`, set the optional `sdkAuthentication` property to `true`.
```javascript
appboy.initialize("YOUR-API-KEY-HERE", {
  baseUrl: "YOUR-SDK-ENDPOINT-HERE",
  sdkAuthentication: true,
});
```
{% endtab %}
{% tab Java %}
When configuring the Appboy instance, call `setIsSdkAuthenticationEnabled` to `true`.
```java
AppboyConfig.Builder appboyConfigBuilder = new AppboyConfig.Builder()
    .setIsSdkAuthenticationEnabled(true);
Appboy.configure(this, appboyConfigBuilder.build());
```
{% endtab %}
{% tab KOTLIN %}
When configuring the Appboy instance, call setIsSdkAuthenticationEnabled to true.
```kotlin
AppboyConfig.Builder appboyConfigBuilder = AppboyConfig.Builder()
    .setIsSdkAuthenticationEnabled(true)
Appboy.configure(this, appboyConfigBuilder.build())
```
{% endtab %}
{% tab Objective-C %}

> Set the `ABKEnableSDKAuthenticationKey` option set to `@YES`. Alternatively, this can be added to your `plist` options using `EnableSDKAuthentication`.

Author: ```suggestion
```

```javascript
appboy.initialize("YOUR-API-KEY-HERE", {
  baseUrl: "YOUR-SDK-ENDPOINT-HERE",
  sdkAuthentication: true,
});
```
{% endtab %}
{% tab Java %}
When configuring the Appboy instance, call `setIsSdkAuthenticationEnabled` to `true`.
```java
AppboyConfig.Builder appboyConfigBuilder = new AppboyConfig.Builder()
    .setIsSdkAuthenticationEnabled(true);
Appboy.configure(this, appboyConfigBuilder.build());
```
{% endtab %}
{% tab KOTLIN %}
When configuring the Appboy instance, call setIsSdkAuthenticationEnabled to true.
```kotlin
AppboyConfig.Builder appboyConfigBuilder = AppboyConfig.Builder()
    .setIsSdkAuthenticationEnabled(true)
Appboy.configure(this, appboyConfigBuilder.build())
```
{% endtab %}
{% tab Objective-C %}
To enable SDK Authentication, add the key `EnableSDKAuthentication` to the `Braze` dictionary in your `.plist` file and set it to true.

Alternatively, you can enable SDK Authentication when initializing the SDK:

```objc
[Appboy startWithApiKey:@"YOUR-API-KEY"
            inApplication:application
        withLaunchOptions:launchOptions
        withAppboyOptions:@{ABKEnableSDKAuthenticationKey : @YES}];
```
{% endtab %}
{% tab Swift %}

> Set the `ABKEnableSDKAuthenticationKey` option set to `@YES`. Alternatively, this can be added to your `plist` options using `EnableSDKAuthentication`.

Author: ```suggestion
To enable SDK Authentication, add the key `EnableSDKAuthentication` to the `Braze` dictionary in your `.plist` file and set it to true.

Alternatively, you can enable SDK Authentication when initializing the SDK:
```

{% sdk_min_versions web:3.4.0 ios:4.2.0 android:14.0.0 %}

### Enable this feature in the Braze SDK.

When this feature is enabled, the Braze SDK will append the current user's last known JWT to network requests made to Braze Servers.

{% alert note %}
Don't worry, initializing with this option alone won't impact data collection in any way, until you start [enforcing authentication](#braze-dashboard) within the Braze Dashboard.
{% endalert %}

{% tabs %}
{% tab Javascript %}
When calling `appboy.initialize`, set the optional `sdkAuthentication` property to `true`.
```javascript
appboy.initialize("YOUR-API-KEY-HERE", {
  baseUrl: "YOUR-SDK-ENDPOINT-HERE",
  sdkAuthentication: true,
});
```
{% endtab %}
{% tab Java %}
When configuring the Appboy instance, call `setIsSdkAuthenticationEnabled` to `true`.
```java
AppboyConfig.Builder appboyConfigBuilder = new AppboyConfig.Builder()
    .setIsSdkAuthenticationEnabled(true);
Appboy.configure(this, appboyConfigBuilder.build());
```
{% endtab %}
{% tab KOTLIN %}

> When configuring the Appboy instance, call setIsSdkAuthenticationEnabled to true.

Author: ```suggestion
When configuring the Appboy instance, call `setIsSdkAuthenticationEnabled` to `true`.
```

```
{% endtab %}
{% endtabs %}

### Set the current user's JWT Token

Whenever your app calls the Braze `changeUser` method, also supply the JWT token that was [generated server-side][4].

You can also update the token to refresh the token mid-session for the current user.

{% alert note %}
Keep in mind that `changeUser` should only be called when the User ID has _actually changed_. You should not use this method as a way to update the signature if the user ID has not changed.
{% endalert %}

{% tabs %}
{% tab Javascript %}
Supply the JWT Token when calling `appboy.changeUser`:

```javascript
appboy.changeUser("NEW-USER-ID", "JWT-TOKEN-FROM-SERVER");
```

Or, when you have refreshed the user's token mid-session:

```javascript
appboy.setAuthenticationToken("NEW-JWT-TOKEN-FROM-SERVER");
```
{% endtab %}
{% tab Java %}

Supply the JWT Token when calling `appboy.changeUser`:

```java
Appboy.getInstance(this).changeUser("NEW-USER-ID", "JWT-TOKEN-FROM-SERVER");
```

Or, when you have refreshed the user's token mid-session:

```java
Appboy.getInstance(this).setSdkAuthenticationSignature("NEW-JWT-TOKEN-FROM-SERVER");
```
{% endtab %}
{% tab KOTLIN %}

> Supply the JWT Token when calling appboy.changeUser:

Author: ```suggestion
Supply the JWT Token when calling `appboy.changeUser`:
```
